subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_d3d.c

```diff
@@ -74,6 +88,62 @@ gst_msdk_frame_get_hdl (mfxHDL pthis, mfxMemId mid, mfxHDL * hdl)
   return MFX_ERR_NONE;
 }
 
+GstMsdkSurface *
+gst_msdk_import_to_msdk_surface (GstBuffer * buf, GstMsdkContext * msdk_context,
+    GstVideoInfo * vinfo, guint flags)
+{
+  GstMemory *mem = NULL;
+  mfxFrameInfo frame_info = { 0, };
+  GstMsdkSurface *msdk_surface = NULL;
+  mfxFrameSurface1 *mfx_surface = NULL;
+  GstMsdkMemoryID *msdk_mid = NULL;
+  GstMapInfo map_info;
+
+  mem = gst_buffer_peek_memory (buf, 0);
+  msdk_surface = g_slice_new0 (GstMsdkSurface);
+
+  if (!gst_is_d3d11_memory (mem) || gst_buffer_n_memory (buf) > 1) {
+    /* d3d11 buffer should hold single memory object */
+    g_slice_free (GstMsdkSurface, msdk_surface);
+    return NULL;
+  }
+
+  if (!gst_buffer_map (buf, &map_info, flags | GST_MAP_D3D11)) {
+    GST_ERROR ("Failed to map buffer");
+    g_slice_free (GstMsdkSurface, msdk_surface);
+    return NULL;
+  }
+
+  /* If buffer has qdata pointing to mfxFrameSurface1, directly extract it */
+  if ((mfx_surface = gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (mem),
```

Comment:
Do we should call this before mapping? If already have the surface, no need to map then.

Reply:
I'm afraid that we can't do in that way. gst_buffer_map with GST_MAP_READ | GST_MAP_D3D11 flags is used to upload the data from staging texture to texutre. If you copy from system memory, first it should be copied to staging texture and second from staging texture to texture. If we do not map before getting qdata, it will cause the texture in mfx_surface in qdata not updated. Returning always the same frame as the first time before setting qdata.